google-cloud-clients/google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/MessageDispatcher.java

```diff
@@ -398,6 +423,14 @@ public void nack() {
             @Override
             public void run() {
               try {
+                if (ackHandler.totalExpiration.plusSeconds(messageDeadlineSeconds.get()).isBefore(now())) {
```

Comment:
What happens if the following sequence occurs:

1. Message is received, but processing is delayed beyond expiration time.
2. Message is redelivered to this subscriber, but we don't add it because it's already in pendingMessages.
3. We now go to process the message, but since we are using the original ackHandler, we think it's expired, even though it got redelivered to us and we just didn't update with the new deadline.

Reply:
When the processing is delayed beyond expiration time, L479 will remove the message from `pendingMessages`. I think two things can happen.

1. We remove message from `pendingMessages` before we redeliver. In this case, we'll just add the redelivered message to the queue. We'll skip processing the original message [a] but process the redelivered one.

2. We get redelivered message before we remove message from `pendingMessages`. This should be rare: if the message is in the `pendingMessages`, we should be keeping that message alive. If it happens though, we won't work on either messages. Eventually the message expires and get delivered for the 3rd time. We work on that one.

[a] We use `messageDeadlineSeconds` as a grace period in case user sets total expiration to 0. It's possible that we remove messgae from `pendingMessages`, then the message gets redelivered before the grace period is over. In that case, we still work on both messages.